System/FilePath/Data/ByteString/Short/Word16.hs

```diff
@@ -0,0 +1,856 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-# OPTIONS_GHC -fno-warn-name-shadowing -fexpose-all-unfoldings #-}
+
+-- |
+-- Module      :  Data.ByteString.Short.Word16
+-- Copyright   :  © 2022 Julian Ospald
+-- License     :  MIT
+--
+-- Maintainer  :  Julian Ospald <hasufell@posteo.de>
+-- Stability   :  experimental
+-- Portability :  portable
+--
+-- ShortByteStrings encoded as UTF16-LE, suitable for windows FFI calls.
+--
+-- Word16s are *always* in BE encoding (both input and output), so e.g. 'pack'
+-- takes a list of BE encoded @[Word16]@ and produces a UTF16-LE encoded ShortByteString.
+--
+-- Likewise, 'unpack' takes a UTF16-LE encoded ShortByteString and produces a list of BE encoded @[Word16]@.
+--
+-- Indices and lengths are always in respect to Word16, not Word8.
+--
+-- All functions will error out if the input string is not a valid UTF16 stream (uneven number of bytes).
```

Comment:
How about a newtype `ShortWord16String` or `Word16String`, enforcing an invariant that number of bytes is even? Otherwise it strikes me as a very unsafe API.

Reply:
Yes, it is unsafe API. The newtype is `AbstractFilePath`.

Reply:
Users are not really meant to deal with this module directly, unless they are building certain primitives based on `AbstractFilePath`. We could rename this module, though. I'd avoid adding another type, we already have 2 newtypes:

1. `OsString`: https://github.com/hasufell/abstract-filepath/blob/a913d715d8971224b45ef22a7b9ceaa263a77601/abstract-filepath-types/lib/AFP/OsString/Internal/Types.hs#L130
2. `WindowsString`: https://github.com/hasufell/abstract-filepath/blob/a913d715d8971224b45ef22a7b9ceaa263a77601/abstract-filepath-types/lib/AFP/OsString/Internal/Types.hs#L65

The invariant would probably be ensured via smart constructor in `WindowsString`.

Reply:
I see, makes sense.

Reply:
It would probably help to have an equivalent of `isValidUTF16` like `isValidUTF8`, but I don't know enough about those cbits: https://github.com/haskell/bytestring/blob/master/cbits/is-valid-utf8.c

Reply:
@kozross

Reply:
@hasufell `isValidUtf16` would require a _very_ different implementation. The techniques used in those cbits do not translate at all.